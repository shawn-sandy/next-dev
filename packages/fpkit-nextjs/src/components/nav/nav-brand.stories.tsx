import { StoryObj, Meta } from "@storybook/react";

import { within } from "@storybook/testing-library";
import { expect } from "@storybook/jest";

// ðŸ‘‰ import your component here - import COMPONENT from "./COMPONENT";
// ðŸ‘‰ replace all instance of COMPONENT with your component name;
// ðŸ‘‰ modify the args to match your component props;
// ðŸ‘‰ remove these comments before committing;

import NavBrand from "./nav-brand";

const meta: Meta<typeof NavBrand> = {
  title: "Nextra Components/NavBrand",
  component: NavBrand,
  // @ts-ignore
  //   subcomponents: { NavItem },
  args: {},
};

export default meta;
type Story = StoryObj<typeof NavBrand>;

const brand = {
  logo: "FPKit NextJS",
  url: "/",
};

export const TextBrand: Story = {
  args: brand,
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const link = canvas.getByRole("link");
    const logo = canvas.getByText(brand.logo);

    // @ts-ignore
    expect(link).toBeInTheDocument();
    // @ts-ignore
    expect(logo).toBeInTheDocument();
  },
};
