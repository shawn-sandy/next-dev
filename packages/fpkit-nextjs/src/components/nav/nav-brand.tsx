import React from "react";

import { Tag, Title, Link } from "@fpkit/react";

export interface BrandTypes {
  url: string;
  logo: React.ReactNode;
}

export const NavBrand = ({ url, logo }: BrandTypes) => {
  return (
    <Tag as="div">
      <Title elm="h1" styles={{ "--fs": "1.2rem" }}>
        <Link href={url} styles={{ textTransform: "capitalize" }}>
          <>{logo}</>
        </Link>
      </Title>
    </Tag>
  );
};

export default NavBrand;
