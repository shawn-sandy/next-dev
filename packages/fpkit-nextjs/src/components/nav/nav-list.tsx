import { MdxFile, FrontMatter } from "nextra";
import { Tag, Link } from "@fpkit/react";
import React from "react";

export type NavListTypes = {
  pages?: MdxFile[];
  styles?: {};
} & FrontMatter;

export const NavList = ({ pages, styles, ...props }: NavListTypes) => {
  if (!pages) return null;
  return (
    <Tag as="ul" styles={styles} key={React.useId()} {...props}>
      {pages?.map((item) => {
        const { route } = item;
        return (
          <Tag as="li" key={React.useId()}>
            <Link href={route}>{item.frontMatter?.title}</Link>
          </Tag>
        );
      })}
    </Tag>
  );
};

export default NavList;
NavList.displayName = "NavList";
