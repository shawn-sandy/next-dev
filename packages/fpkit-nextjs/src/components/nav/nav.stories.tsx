import { StoryObj, Meta } from "@storybook/react";

import { within } from "@storybook/testing-library";
import { expect } from "@storybook/jest";

import Nav, { NavItem } from "./nav";

const meta: Meta<typeof Nav> = {
  title: "Nextra Components/Nav",
  component: Nav,
  // @ts-ignore
  subcomponents: { NavItem },
  argTypes: {
    children: {
      control: { type: "object" },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Nav>;

const navList = (
  <ul>
    <li>
      <a href="/about">About</a>
    </li>
    <li>
      <a href="/blog">Blog</a>
    </li>
    <li>
      <a href="/contact">Contact</a>
    </li>
  </ul>
);

export const BaseNav: Story = {
  args: {
    children: navList,
  },
};

const navListItems = (
  <ul>
    <NavItem route="/about">About</NavItem>
    <NavItem route="/blog">Blog</NavItem>
    <NavItem route="/contact">Contact</NavItem>
  </ul>
);

export const NavItems: Story = {
  args: {
    children: navListItems,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const aboutLink = canvas.getByRole("link", { name: "About" });
    const links = canvas.getAllByRole("link");
    expect(links).toHaveLength(3);
  },
};
