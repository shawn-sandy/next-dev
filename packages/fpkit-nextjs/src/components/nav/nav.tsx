import React from "react";

import { Tag } from "@fpkit/react";

export type NavItemProps = {
  route: string;
  styles?: {};
  children?: React.ReactNode;
};

/**
 * NavItem component creates a navigation link element wrapped in a list item `li`
 * @param param0
 * @returns
 */
export const NavItem = ({
  route,
  styles,
  children,
  ...props
}: NavItemProps) => {
  return (
    <>
      <Tag as="li" styles={styles} {...props}>
        <Tag as="a" href={route} styles={{ textTransform: "capitalize" }}>
          <>{children}</>
        </Tag>
      </Tag>
    </>
  );
};

export type NavProps = {
  children?: React.ReactNode;
  styles?: {};
};

/**
 * Nav component creates a navigation element
 * @param param0
 * @returns
 */
export const Nav = ({ children, styles, ...props }: NavProps) => {
  return (
    <>
      <Tag as="nav" style={styles} {...props}>
        <>{children}</>
      </Tag>
    </>
  );
};

/**
 * Exports
 */
export default Nav;
Nav.displayName = "Nav";
