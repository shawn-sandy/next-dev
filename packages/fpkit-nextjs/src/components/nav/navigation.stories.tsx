import { StoryObj, Meta } from "@storybook/react";

import { within, screen } from "@storybook/testing-library";
import { expect } from "@storybook/jest";

// ðŸ‘‡ import you component here -- import Component form "./component";
import Navigation from "./navigation";

// replace Component with your component name
const meta: Meta<typeof Navigation> = {
  title: "Nextra Components/Navigation",
  // @ts-ignore
  component: Navigation,
  // @ts-ignore
  argTypes: {},
};

const dataSet = [
  {
    kind: "Meta",
    data: {
      index: "Introduction",
      another: "Another Page",
      advanced: "Advanced (A Folder)",
      about: {
        title: "About",
        type: "page",
      },
      contact: {
        title: "Contact â†—",
        type: "page",
        href: "https://twitter.com/shuding_",
        newWindow: true,
      },
    },
  },
  {
    kind: "MdxPage",
    name: "about",
    route: "/about",
  },
  {
    kind: "Folder",
    name: "advanced",
    route: "/advanced",
    children: [
      {
        kind: "MdxPage",
        name: "satori",
        route: "/advanced/satori",
      },
      {
        kind: "Meta",
        data: {
          satori: "Satori",
        },
      },
    ],
  },
  {
    kind: "MdxPage",
    name: "advanced",
    route: "/advanced",
    frontMatter: {
      title: "Advanced title",
      description:
        "The city bustled with people rushing to work, the sound of honking cars and chatter filling the air. The tall skyscrapers towered over the streets, their windows glinting in the sunlight. A sense of hustle and bustle prevailed in this concrete jungle.",
      date: "2023-01-22T23:51:45.574Z",
    },
  },
  {
    kind: "MdxPage",
    name: "another",
    route: "/another",
    frontMatter: {
      title: "Another page",
      description:
        "Web developers work tirelessly to create seamless digital experiences. They code and debug, troubleshoot and optimize, all to deliver optimal performance and user satisfaction.",
    },
  },
  {
    kind: "MdxPage",
    name: "index",
    route: "/",
    frontMatter: {
      title: "Introduction",
      description:
        "A simple, powerful and flexible site generation framework with everything you love from Next.js.",
    },
  },
];

const branding = {
  url: "/",
  logo: "Branding Logo",
};

export default meta;
type Story = StoryObj<typeof Navigation>;

export const NavigationBar: Story = {
  args: {
    // @ts-ignore
    data: dataSet,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const nav = canvas.getByRole("navigation");
    expect(nav).toBeDefined();
    const links = canvas.getByText(branding.logo);
    expect(links).toBeDefined();
  },
};

export const BrandedNavigationBar: Story = {
  args: {
    ...NavigationBar.args,
    // @ts-ignore
    brand: branding,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const nav = canvas.getByRole("navigation");
    const brand = canvas.getByRole("heading");
    const links = canvas.getByText(branding.logo);

    expect(nav).toBeDefined();
    expect(brand).toBeDefined();
    expect(brand.getElementsByTagName("a")).toBeDefined();
    expect(links).toBeDefined();
  },
};
