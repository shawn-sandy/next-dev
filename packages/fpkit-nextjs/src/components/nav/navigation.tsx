import React from "react";
import { Tag, Title } from "@fpkit/react";
import Nav, { NavItem } from "./nav";

import { MdxFile, Folder, PageMapItem } from "nextra";

type Kind = "Meta" | "MdxPage" | "Folder";

interface Meta {
  index: string;
  another: string;
  advanced: string;
  about: {
    title: string;
    type: string;
  };
  contact: {
    title: string;
    type: string;
    href: string;
    newWindow: boolean;
  };
  kind: "Meta";
}

type Data = (Meta | MdxFile | Folder | PageMapItem) & {
  kind: Kind;
};

function getFolders(data: Data[]): Folder[] {
  const folders: Folder[] = [];

  data.forEach((obj) => {
    if (obj.kind === "Folder") {
      folders.push(obj);
    }
  });

  return folders;
}

export interface Brand {
  brand?: {
    url?: string;
    logo?: React.ReactNode;
  };
}

interface NavbarProps extends Brand {
  styles?: {};
  data: Data[];
}

export const Navigation: React.FC<NavbarProps> = ({ data, brand, styles }) => {
  const folders = getFolders(data);

  return (
    <Nav styles={styles}>
      {!!brand && (
        <Tag as="div">
          <Title elm="h1" styles={{ "--fs": "1.2rem" }}>
            <Tag
              as="a"
              href={brand.url}
              styles={{ textTransform: "capitalize" }}
            >
              <>{brand.logo}</>
            </Tag>
          </Title>
        </Tag>
      )}

      <Tag as="ul">
        {folders.map((folder) => (
          <NavItem route={folder.route} key={folder.name}>
            <>{folder.name}</>
          </NavItem>
        ))}
      </Tag>
    </Nav>
  );
};

export default Navigation;
