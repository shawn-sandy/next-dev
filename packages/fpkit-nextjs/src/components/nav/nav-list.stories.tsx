import { StoryObj, Meta } from "@storybook/react";
import React from "react";

import { within } from "@storybook/testing-library";
import { expect } from "@storybook/jest";

// ðŸ‘‡ import you component here -- import Component form "./component";
import NavList from "./nav-list";
import { Nav } from "./nav";
import { navListItems } from "../data/nav-list";

// replace Component with your component name
const meta: Meta<typeof NavList> = {
  title: "Nextra Components/NavList",
  component: NavList,
  // @ts-ignore
  args: {
    pages: navListItems,
  },
};

export default meta;
type Story = StoryObj<typeof NavList>;

export const Navbar: Story = {
  args: {},
  render: (args) => {
    return (
      <Nav>
        <NavList {...args} />
      </Nav>
    );
  },
};

export const NavbarList: Story = {
  args: {},
};
