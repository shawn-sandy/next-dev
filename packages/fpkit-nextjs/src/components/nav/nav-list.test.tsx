import React from "react";

import { describe, expect, test } from "vitest";
import { render, screen } from "@testing-library/react";
import { axe } from "vitest-axe";

import { NavList } from "./nav-list";
import { Nav } from "./nav";

import { navListItems } from "../data/nav-list";

describe("Nav renders correctly", () => {
  test("renders without crashing", async () => {
    render(
      <Nav>
        <NavList pages={navListItems} />
      </Nav>
    );
    const nav = screen.getByRole("navigation");
    const list = screen.getByRole("list");
    const links = screen.getAllByRole("link");

    expect(nav).toBeDefined();
    expect(list).toBeDefined();
    expect(links).toHaveLength(navListItems.length);
    // @ts-ignore
    expect(await axe(nav)).toHaveNoViolations();
    expect(nav).toMatchSnapshot();
  });
});
