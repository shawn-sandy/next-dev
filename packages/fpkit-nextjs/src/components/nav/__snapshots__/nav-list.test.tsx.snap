// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nav renders correctly > renders without crashing 1`] = `
<nav>
  <ul>
    <li>
      <a
        href="/articles"
      >
        Articles
      </a>
    </li>
    <li>
      <a
        href="/"
      >
        About
      </a>
    </li>
    <li>
      <a
        href="/post"
      >
        Posts
      </a>
    </li>
  </ul>
</nav>
`;
